appdev/connectors/mysql_connector.py

```diff
@@ -0,0 +1,105 @@
+from threading import Thread
+from Queue import Queue, Empty
+import mysql.connector
+
+
+class DbThread(Thread):
+  def __init__(self, connection_pool, table, input_queue, result_queue=None,
+               queue_timeout=0.5):
+    super(DbThread, self).__init__()
+    self.connection_pool = connection_pool
+    self.table = table
+    self.input_queue = input_queue
+    self.result_queue = result_queue
+    self.queue_timeout = queue_timeout
+
+
+class WriterThread(DbThread):
+  def run(self):
+    # Create query in the format:
+    #   "INSERT INTO salaries (emp_no, salary, from_date, to_date)
+    #     VALUES (%(emp_no)s, %(salary)s, %(from_date)s, %(to_date)s)"
+    # data is dictionary such as {'emp_no': 1, 'salary': 2, ...}
+    while not self.input_queue.empty():
+      connection = self.connection_pool.get()
+      cursor = connection.cursor()
+      try:
+        # A thread may enter the loop, but another thread could pluck the
+        # last item. This timeout is to prevent a thread from waiting for
+        # another item that will never come.
+        data = self.input_queue.get(timeout=self.queue_timeout)
+      except Empty:
+        break
+      values_string = ['%({})s'.format(col) for col in data.keys()]
+      insert_query = "INSERT INTO {} ({}) VALUES ({})" \
+        .format(self.table, ', '.join(data.keys()),
+                ', '.join(values_string))
+      cursor.execute(insert_query, data)
+      connection.commit()
+      self.input_queue.task_done()
+      self.connection_pool.put(connection)
+
+
+class ReaderThread(DbThread):
+  def run(self):
+    while not self.input_queue.empty():
+      connection = self.connection_pool.get()
+      cursor = connection.cursor()
+      try:
+        a, b = self.input_queue.get(timeout=self.queue_timeout)
+      except Empty:
+        break
+      read_query = "SELECT * FROM {} LIMIT {} OFFSET {}" \
+                     .format(self.table, b-a+1, a)
+      cursor.execute(read_query)
+      result = []
+      for row in cursor:
+        result.append(row)
+      self.input_queue.task_done()
+      self.result_queue.put(result)
+      self.connection_pool.put(connection)
+
+
+class MySQLConnector(object):
+  def __init__(self, user, password, host, database,
+               num_connections=10, num_threads=10):
+    self.config = {
+        'user': user,
+        'password': password,
+        'host': host,
+        'database': database
+    }
+    self.connection_pool = Queue()
+    self.num_threads = num_threads
+    for _ in range(num_connections):
+      cnx = mysql.connector.connect(**self.config)
+      self.connection_pool.put(cnx)
```

Comment:
Because of this, please include a scaleup and scaledown feature of the connection pool to handle the current load of the system which may/may not be peak.